Baikal/Kernels/CL/scene.cl

```diff
@@ -247,53 +265,192 @@ void Scene_FillDifferentialGeometry(// Scene
-        diffgeo->dpdu = normalize(GetOrthoVector(diffgeo->n));
-        diffgeo->dpdv = normalize(cross(diffgeo->n, diffgeo->dpdu));
+        diffgeo->dpdu = diffgeo->dpdv = 0.0f;
+        diffgeo->dndu = diffgeo->dndv = 0.0f;
     }
-}
 
+    // Initialize screen space uv derivatives
+    diffgeo->duvdx = diffgeo->duvdy = 0.0f;
+
+    // Compute tangent and bitangent vectors
+    diffgeo->tangent = diffgeo->dpdu;
+    // Gram–Schmidt orthogonalization
+    diffgeo->tangent -= dot(diffgeo->n, diffgeo->tangent) * diffgeo->n;
+    diffgeo->tangent = normalize(diffgeo->tangent);
+    // Bitangent vector need to be orthogonal to tangent, since we use
+    // the advantage of an orthogonal matrix that is easily invertible
+    diffgeo->bitangent = normalize(cross(diffgeo->tangent, diffgeo->n));
+/*
+    // Check handedness when uv are mirrored
+    if (dot(cross(diffgeo->n, diffgeo->tangent), diffgeo->bitangent) < 0.0f)
+    {
+        diffgeo->tangent *= -1.0f;
+    }
+*/
+
+
+}
 
 // Calculate tangent transform matrices inside differential geometry
 INLINE void DifferentialGeometry_CalculateTangentTransforms(DifferentialGeometry* diffgeo)
 {
-    diffgeo->world_to_tangent = matrix_from_rows3(diffgeo->dpdu, diffgeo->n, diffgeo->dpdv);
+    diffgeo->world_to_tangent = matrix_from_rows3(diffgeo->tangent, diffgeo->n, diffgeo->bitangent);
 
-    diffgeo->world_to_tangent.m0.w = -dot(diffgeo->dpdu, diffgeo->p);
+    diffgeo->world_to_tangent.m0.w = -dot(diffgeo->tangent, diffgeo->p);
     diffgeo->world_to_tangent.m1.w = -dot(diffgeo->n, diffgeo->p);
-    diffgeo->world_to_tangent.m2.w = -dot(diffgeo->dpdv, diffgeo->p);
+    diffgeo->world_to_tangent.m2.w = -dot(diffgeo->bitangent, diffgeo->p);
 
-    diffgeo->tangent_to_world = matrix_from_cols3(diffgeo->world_to_tangent.m0.xyz, 
+    diffgeo->tangent_to_world = matrix_from_cols3(diffgeo->world_to_tangent.m0.xyz,
         diffgeo->world_to_tangent.m1.xyz, diffgeo->world_to_tangent.m2.xyz);
 
     diffgeo->tangent_to_world.m0.w = diffgeo->p.x;
     diffgeo->tangent_to_world.m1.w = diffgeo->p.y;
     diffgeo->tangent_to_world.m2.w = diffgeo->p.z;
 }
 
+// Compute screen-space derivative of uv coords
+INLINE void DifferentialGeometry_CalculatePartialDerivatives(
+                          // Auxiliary ray
+                          GLOBAL aux_ray const* my_ray,
+                          // Differential geometry
+                          DifferentialGeometry* diffgeo,
+                          float3* position_derivative,
+                          float2* uv_derivative
+                          )
+{
+    //float3 o = vload_half3(0, (GLOBAL half*)&my_ray->o);
+    //float3 d = vload_half3(0, (GLOBAL half*)&my_ray->d);
+    float3 o = my_ray->o;
+    float3 d = my_ray->d;
+
+    // Find intersection point of auxiliary ray with the dpdu plane
+    float t = dot(diffgeo->n, diffgeo->p - o) / dot(diffgeo->n, d);
+
+    float3 p = o + d * t;
+
+    // Next we need to find uv offset from the position offset
```

Comment:
remove dead code

Reply:
This is the comment with the formula, not a line of code